design/observerability-2.md

```diff
@@ -0,0 +1,196 @@
+- Prometheus - A time-series database. We store metrics there.
+- Grafana - A service that presents diagrams based on metrics from Prometheus.
+- EFK Stack - Elsticsearch, Fluentd and Kibana. A typical log-collecting stack.
+- CLO - Cluster Logging Operator
+- CR - Custom Resource. An API provided by Keycloak Operator for the context of this document.
+
+## Motivation
+
+Managing Keycloak cluster requires involving a dedicated SRE (Site Reliability Engineering) Team. Such a team needs to
+be able to:
+
+* Diagnose if Keycloak cluster is up and running
+* Have sufficient information to diagnose issues
+* Have enough data and metrics to foresee problems that will happen in the near future
+
+Typically, an SRE Team uses dashboards and alerts. Both of them are created based on metrics exported by a Service
+(in this case, the Service is both Keycloak and Keycloak Operator).
+
+Based on metrics, dashboards and alerts, it is possible to define SLA - an agreement between a Service Provider and a
+Service Consumer.
+
+## Scope
+
+What's in the scope of this design:
+- Designing a new Healcheck API. This has been covered in a separate design document. See [KEYCLOAK-12398](https://issues.redhat.com/browse/KEYCLOAK-12398)
+- Designing and implementing a dashboard for the SRE Team.
+- Defining SLIs and SLOs as well as proposing template SLAs for a managed Keycloak cluster.
+- Defining metrics required by the proposed dashboard as well as designing a mechanism for exporting them to prometheus.
+- Design logging improvements for better error tracking.
+
+What's out of the scope:
+- Traceability - tracking a user across the whole authorization flow. At the moment we don't have any requirements
+in this direction. If we get such a requirement, we will track this progress in a separate epic and design.
+
+## SRE Dashboard
+
+![SRE Dashboard](img/observability-dashboard.png)
+
+NOTE: The diagram implementation and specific charts might change during the implementation phase. The above proposal
+shows only the general idea.
+
+The proposal consists of 3 sections:
+- System health - Enables the SRE Team to quickly tell if Keycloak cluster is up and running. This section consists of:
+  - Readiness and Liveness Probes chart
+  - HTTP 5xx error rate for all major Keycloak Services, including: Admin Console, Account Console, OIDC, SAML
+  - Aggregated latency across the most important endpoints
+  - (Optional) Error rate in logs [^1]
+- Detailed Metrics - This section helps foresee upcoming failures. It includes:
+  - Pod memory, CPU and container storage charts
+  - GC Times
+  - Heap and non-heap memory metrics
+- Experimental metrics - Acts as a bridge between the SRE and Keycloak Teams. We introduce new metrics here and if they
+are useful, we promote them to other sections. This section includes:
+  - External IdP state (Social Login Providers)
+  - The Total number of CRs discovered by the Operator vs the number of CRs successfully reconciled
```

Comment:
I don't like the idea of experimental metrics. It's just a dumping ground for things that we think may or may not make sense.

I'd say we at least need one more section which is "external services" where external IdPs, user fed, db health, etc would be shown.

Reply:
Total number of CRs discovered by the Operator - this cries to me that there maybe should be a separate dashboard and metrics for the Operator.

Reply:
> I don't like the idea of experimental metrics. It's just a dumping ground for things that we think may or may not make sense.

I believe we need some section that we'll use to introduce new metrics. They might misbehave at first and I wouldn't like anyone to be page based on them. If they live in an area called "Experimental" - the contract seems to be clear. Perhaps you have some suggestion how to improve this?

> Total number of CRs discovered by the Operator - this cries to me that there maybe should be a separate dashboard and metrics for the Operator.

That's actually a good point. There's one more metric we could put there - the liveness/readiness probe. Do you have any other ideas?

Reply:
Introduced another category - "Operator Metrics".